pkg/hubble/exporter/cell.go

```diff
@@ -0,0 +1,170 @@
+	// "/var/run/cilium/hubble/events.log".
+	ExportFilePath string `mapstructure:"hubble-export-file-path"`
+	// ExportFileMaxSizeMB specifies the file size in MB at which to rotate the
+	// Hubble export file.
+	ExportFileMaxSizeMB int `mapstructure:"hubble-export-file-max-size-mb"`
+	// ExportFileMaxBackups specifies the number of rotated files to keep.
+	ExportFileMaxBackups int `mapstructure:"hubble-export-file-max-backups"`
+	// ExportFileCompress specifies whether rotated files are compressed.
+	ExportFileCompress bool `mapstructure:"hubble-export-file-compress"`
+	// ExportAllowlist specifies allow list filter use by exporter.
+	ExportAllowlist []*flowpb.FlowFilter `mapstructure:"hubble-export-allowlist"`
+	// ExportDenylist specifies deny list filter use by exporter.
+	ExportDenylist []*flowpb.FlowFilter `mapstructure:"hubble-export-denylist"`
+	// ExportFieldmask specifies list of fields to log in exporter.
+	ExportFieldmask []string `mapstructure:"hubble-export-fieldmask"`
+}
+
+var defaultConfig = config{
+	FlowlogsConfigFilePath: "",
+	ExportFilePath:         exporteroption.Default.Path,
+	ExportFileMaxSizeMB:    exporteroption.Default.MaxSizeMB,
+	ExportFileMaxBackups:   exporteroption.Default.MaxBackups,
+	ExportFileCompress:     exporteroption.Default.Compress,
+	ExportAllowlist:        []*flowpb.FlowFilter{},
+	ExportDenylist:         []*flowpb.FlowFilter{},
+	ExportFieldmask:        []string{},
+}
+
+func (def config) Flags(flags *pflag.FlagSet) {
+	flags.String("hubble-flowlogs-config-path", def.FlowlogsConfigFilePath, "Filepath with configuration of hubble flowlogs")
+	flags.String("hubble-export-file-path", def.ExportFilePath, "Filepath to write Hubble events to. By specifying `stdout` the flows are logged instead of written to a rotated file.")
+	flags.Int("hubble-export-file-max-size-mb", def.ExportFileMaxSizeMB, "Size in MB at which to rotate Hubble export file.")
+	flags.Int("hubble-export-file-max-backups", def.ExportFileMaxBackups, "Number of rotated Hubble export files to keep.")
+	flags.Bool("hubble-export-file-compress", def.ExportFileCompress, "Compress rotated Hubble export files.")
+	flags.StringSlice("hubble-export-allowlist", []string{}, "Specify allowlist as JSON encoded FlowFilters to Hubble exporter.")
+	flags.StringSlice("hubble-export-denylist", []string{}, "Specify denylist as JSON encoded FlowFilters to Hubble exporter.")
+	flags.StringSlice("hubble-export-fieldmask", def.ExportFieldmask, "Specify list of fields to use for field mask in Hubble exporter.")
+}
+
+func (cfg config) validate() error {
+	if fm := cfg.ExportFieldmask; len(fm) > 0 {
+		_, err := fieldmaskpb.New(&flowpb.Flow{}, fm...)
+		if err != nil {
+			return fmt.Errorf("hubble-export-fieldmask contains invalid fieldmask '%v': %w", fm, err)
+		}
+	}
+	return nil
+}
+
+type hubbleExportersOut struct {
+	cell.Out
+
+	ObserverOptions []observeroption.Option `group:"hubble-observer-options,flatten"`
```

Comment:
I think these are too "small" to be in the object graph individually. In my head, hive is most useful at the component level - tying together a large application with many components. The heuristic for me is "Do I want to see XYZ in the `cilium hive` output?". With individual options for a component, I'd lean towards no. (The best docs for hive and its use in cilium is likely https://docs.cilium.io/en/latest/contributing/development/hive/#id1)

Not sure it needs to be part of this PR, and I wouldn't want to block this work on it, but you could potentially wrap these up into a single config/options struct with a slice and then `ProvidePrivate` it from the toplevel hubble module to just the hubble components? It depends on who needs to access this, but I'm guessing it's not the whole, global object graph.

Reply:
> hi, sorry for the drive-by review ✌️
>
> Maybe it's worth requesting a review from the foundations team when adding new cells :)

Thank you for the review, its very much appreciated, and agreed, will do that going forward!

If I understand correctly, your suggestion is to not create a Cell for the hubble exporters and instead provide these newX constructors directly on the Hubble Cell (and maybe as only one constructor)? Can we also declare multiple configuration as part of one cell? Oh and I assumed that a Cell declared into a parent Cell is only available to parent and child. I guess that is wrong?

Going forward, the Hubble cell will be broken up into 10-13 logical components. Wouldn't it make more sense for these to be Cells instead of plain providers for organization/config declaration purposes? The main idea behind this work is to break up this gigantic Hubble cell to make it easier to work with and isolate/declare dependencies. I feel if we declare everything on one Cell with one single configuration/metrics/etc, we would not gain much from the current setup, but I might be misunderstanding your comment.

One thing that might not be clear is we are providing these exporters as options because we want to hook ourselves into the hubble observer service as an `OnDecodedEvent` hook, but these are fully fledge components, and they could also be provided as `FlowExporter`. It depends if we want Hubble to be aware of "exporters" or only of the hooks.

Happy to discuss further our use-case and find the best approach. I'll also CC @kaworu to get his input on this.

Thanks again for jumping in and helping us 😸

Reply:
Hey, sorry for the delay

> Can we also declare multiple configuration as part of one cell?

Yeah, I wouldn't know why this would not work. `pkg/clustermesh/cell.go` does it too, for example.

> I also assumed that a Cell declared into a parent Cell is only available to parent and child, but I guess that is wrong?

If I understand correctly, yes, that's not true. Declaring a cell doesn't do much, but `Provide` is the key. What you `Provide` to the object graph is global, unless you use `ProvidePrivate`. Think of `Provide` as the way to declare the interface/service you're offering to the whole application. (To ensure I wasn't spreading misinformation I just edited the example in github.com/cilium/hive, `provide`d an `int` from a submodule and `invoke`d a `func(int)` top-level - it gets the `int`.)

I don't have enough insight in this area of the codebase to give well-founded answers to your questions, I'm afraid. I don't have a strong opinion on whether hubble should be a single cell or a few, as long as each cell either has a sensible API to offer to the rest of the application, or is private to only the hubble module.

If you run `go run ./daemon hive` you can look at some of the existing modules/cells, and see which parts of the are private (the :lock: emoji) to see some examples.

Reply:
Thanks for the details, I think I understand better what you mean and I agree that its good to keep dependencies encapsulated when possible.
I went ahead and pushed a new commit that removes the Hubble-exporters cell and instead ProvidePrivate what we need directly on the hubble cell.